File: backend/app/models/pool_template.py
"""PoolTemplate domain model for reusable pool configurations.

The PoolTemplate model stores pre-defined and custom pool structures
that can be applied to properties to quickly set up expense pool hierarchies.
"""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


class PoolStructureNode(BaseModel):
    """Single node in a pool structure hierarchy.

    Represents a pool with optional nested children pools.
    """

    name: str = Field(..., min_length=1, max_length=100)
    gross_up_enabled: bool = Field(default=True)
    children: list["PoolStructureNode"] = Field(default_factory=list)

    @field_validator("children")
    @classmethod
    def validate_max_depth(
        cls, children: list["PoolStructureNode"]
    ) -> list["PoolStructureNode"]:
        """Validate maximum depth of 2 levels (parent → child only)."""
        if children:
            for child in children:
                if child.children:
                    raise ValueError(
                        "Pool hierarchy cannot exceed 2 levels (parent → child only)"
                    )
        return children


class PoolTemplateStructure(BaseModel):
    """Structure of a pool template.

    Contains the pool hierarchy as a list of root-level pools
    with optional nested children.
    """

    pools: list[PoolStructureNode]

    @field_validator("pools")
    @classmethod
    def validate_non_empty(
        cls, pools: list[PoolStructureNode]
    ) -> list[PoolStructureNode]:
        """Validate at least one pool exists."""
        if not pools:
            raise ValueError("Template must contain at least one pool")
        return pools


class PoolTemplate(BaseModel):
    """Full pool template model from database.

    Pool templates allow users to apply pre-defined pool structures
    to properties, saving time when setting up new properties.
    """

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    property_type: str | None = Field(
        None, description="Property type filter: retail, office, mixed_use, etc."
    )
    structure: dict[str, Any] = Field(
        ..., description="Pool hierarchy as JSON structure"
    )
    is_system: bool = Field(
        default=False,
        description="True for system templates, False for custom templates",
    )
    organization_id: UUID | None = Field(
        None,
        description="NULL for system templates, organization ID for custom templates",
    )
    version: int = Field(default=1, ge=1)
    created_at: datetime
    updated_at: datetime

    @field_validator("structure")
    @classmethod
    def validate_structure(cls, structure: dict[str, Any]) -> dict[str, Any]:
        """Validate structure matches PoolTemplateStructure schema."""
        # Parse and validate structure
        PoolTemplateStructure(**structure)
        return structure


class PoolTemplateCreate(BaseModel):
    """DTO for creating a custom pool template.

    Only for creating custom (non-system) templates.
    Organization ID will be set from authenticated user context.
    """

    name: str = Field(..., min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    property_type: str | None = None
    structure: dict[str, Any]

    @field_validator("structure")
    @classmethod
    def validate_structure(cls, structure: dict[str, Any]) -> dict[str, Any]:
        """Validate structure matches PoolTemplateStructure schema."""
        PoolTemplateStructure(**structure)
        return structure


class PoolTemplateUpdate(BaseModel):
    """DTO for updating a custom pool template.

    System templates cannot be updated.
    """

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    property_type: str | None = None
    structure: dict[str, Any] | None = None

    @field_validator("structure")
    @classmethod
    def validate_structure(
        cls, structure: dict[str, Any] | None
    ) -> dict[str, Any] | None:
        """Validate structure matches PoolTemplateStructure schema if provided."""
        if structure is not None:
            PoolTemplateStructure(**structure)
        return structure


class PoolTemplateList(BaseModel):
    """Lightweight template for listing purposes."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    property_type: str | None
    is_system: bool
    pool_count: int = Field(
        default=0, description="Number of root-level pools in template"
    )
    created_at: datetime


class ApplyTemplateRequest(BaseModel):
    """Request to apply a template to a property."""

    template_id: UUID = Field(..., description="ID of template to apply")
    property_id: UUID = Field(..., description="ID of property to apply template to")
    delete_existing: bool = Field(
        default=True,
        description="Whether to delete existing pools before applying template",
    )

File: frontend/src/types/pool-template.ts
/**
 * Pool template types for reusable pool configurations.
 *
 * Pool templates allow users to quickly set up expense pool structures
 * by applying pre-defined or custom templates to properties.
 */

import { z } from 'zod'

/**
 * Single node in pool hierarchy (parent or child pool).
 */
type PoolStructureNodeInput = {
  name: string
  gross_up_enabled?: boolean
  children?: PoolStructureNodeInput[]
}

export type PoolStructureNode = {
  name: string
  gross_up_enabled: boolean
  children: PoolStructureNode[]
}

export const PoolStructureNodeSchema: z.ZodType<
  PoolStructureNode,
  z.ZodTypeDef,
  PoolStructureNodeInput
> = z.object({
  name: z.string().min(1).max(100),
  gross_up_enabled: z.boolean().default(true),
  children: z
    .array(z.lazy(() => PoolStructureNodeSchema))
    .default([]) as z.ZodDefault<
    z.ZodArray<
      z.ZodLazy<
        z.ZodType<PoolStructureNode, z.ZodTypeDef, PoolStructureNodeInput>
      >
    >
  >,
}) as z.ZodType<PoolStructureNode, z.ZodTypeDef, PoolStructureNodeInput>

/**
 * Complete pool template structure with validation.
 */
export const PoolTemplateStructureSchema = z.object({
  pools: z.array(PoolStructureNodeSchema).min(1),
})

export type PoolTemplateStructure = z.infer<typeof PoolTemplateStructureSchema>

/**
 * Full pool template from database.
 */
export const PoolTemplateSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  property_type: z.string().nullable(),
  structure: z.record(z.any()), // JSONB structure
  is_system: z.boolean(),
  organization_id: z.string().uuid().nullable(),
  version: z.number().int().min(1),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
})

export type PoolTemplate = z.infer<typeof PoolTemplateSchema>

/**
 * Lightweight template for listing.
 */
export const PoolTemplateListSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  description: z.string().nullable(),
  property_type: z.string().nullable(),
  is_system: z.boolean(),
  pool_count: z.number().int().min(0).default(0),
  created_at: z.string().datetime(),
})

export type PoolTemplateList = z.infer<typeof PoolTemplateListSchema>

/**
 * DTO for creating custom pool template.
 */
export const PoolTemplateCreateSchema = z.object({
  name: z.string().min(1).max(100),
  description: z.string().max(500).nullable().optional(),
  property_type: z.string().nullable().optional(),
  structure: z.record(z.any()), // Will be validated on backend
})

export type PoolTemplateCreate = z.infer<typeof PoolTemplateCreateSchema>

/**
 * DTO for updating custom pool template.
 */
export const PoolTemplateUpdateSchema = z.object({
  name: z.string().min(1).max(100).optional(),
  description: z.string().max(500).nullable().optional(),
  property_type: z.string().nullable().optional(),
  structure: z.record(z.any()).optional(),
})

export type PoolTemplateUpdate = z.infer<typeof PoolTemplateUpdateSchema>

/**
 * Request to apply template to property.
 */
export const ApplyTemplateRequestSchema = z.object({
  template_id: z.string().uuid(),
  property_id: z.string().uuid(),
  delete_existing: z.boolean().default(true),
})

export type ApplyTemplateRequest = z.infer<typeof ApplyTemplateRequestSchema>

/**
 * Response from applying template to property.
 */
export const ApplyTemplateResponseSchema = z.object({
  template_name: z.string(),
  property_id: z.string().uuid(),
  pools_created: z.number().int().min(0),
  parent_pools: z.array(z.record(z.any())),
  child_pools: z.array(z.record(z.any())),
})

export type ApplyTemplateResponse = z.infer<typeof ApplyTemplateResponseSchema>
